### PR TITLE
Achievement: Goal System Rework

### PR DESCRIPTION
**Pull request by jetkan-yk**
### Description

Refactor goal system to meet the new backend goal requirements.

Please refer to [this Google Doc](https://docs.google.com/document/d/1coFLbRhNiWL7MK1bbv8XSgxPkmpDWnb8f0t-rTQMRX8/edit#) for details.

### Upcoming Changes

The next PR will focus on refactoring the Achievement Control and all its related subcomponents, including these files that are commented out intentionally.

https://github.com/source-academy/cadet-frontend/blob/2709489c5607f18fd3d0db0c6bdbf821dfe3e655/src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx#L1-L65

https://github.com/source-academy/cadet-frontend/blob/2709489c5607f18fd3d0db0c6bdbf821dfe3e655/src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoals.tsx#L1-L99

### Important

The following lines in `GameUserStateManager` require changes after the backend is updated.

Created an issue as a reminder: #1438
cc: @anthony-halim @sciffany
https://github.com/source-academy/cadet-frontend/blob/0204b1fa9936fc59b8cd6584bfa1bbd4b2b4a449/src/features/game/state/GameUserStateManager.ts#L91-L97

### CODE REVIEW

File: src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx [hunk 1]
```diff
@@ -1,3 +1,10 @@
+import React from 'react';
+
+function EditableAchievementGoal() {
+  return <></>;
+}
```
Comment (anthony-halim) [inline src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx:5]: What is this for?

File: src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoals.tsx [hunk 2]
```diff
@@ -1,17 +1,21 @@
-import { Button, Dialog } from '@blueprintjs/core';
-import React, { useState } from 'react';
+import React from 'react';
 
+function EditableAchievementGoals() {
+  return <></>;
+}
```
Comment (anthony-halim) [inline src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoals.tsx:5]: Same as the above

File: src/commons/achievement/control/editorTools/EditableAchievementCard.tsx [hunk 3]
```diff
@@ -108,10 +105,11 @@ function EditableAchievementCard(props: EditableAchievementCardProps) {
     setUnsaved();
   };
 
+  /*
   const handleEditGoals = (goals: AchievementGoal[], shouldUpdate: boolean) => {
     setEditableAchievement({
       ...editableAchievement,
```
Comment (anthony-halim) [inline src/commons/achievement/control/editorTools/EditableAchievementCard.tsx:111]: Remove unused code

File: src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx [hunk 4]
```diff
@@ -1,3 +1,10 @@
...
+
+/*
```
Comment (anthony-halim) [inline src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx:7]: Remove unused code

File: src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoals.tsx [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
+
+/*
```
Comment (anthony-halim) [inline src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoals.tsx:7]: Remove unused code

File: src/commons/achievement/utils/DateHelper.ts [hunk 6]
```diff
@@ -1,4 +1,13 @@
-export const prettifyDate = (deadline: Date) => {
+const now = new Date();
```
Comment (anthony-halim) [inline src/commons/achievement/utils/DateHelper.ts:1]: The server can run for more than a day.

File: src/features/game/state/GameUserStateManager.ts [hunk 7]
```diff
@@ -86,11 +88,18 @@ export default class GameUserStateManager {
    */
   public async loadAchievements() {
     const achievements = SourceAcademyGame.getInstance().getAchievements();
+    // TODO: replace below with const goals = SourceAcademyGame.getInstance().getGoals();
+    // This is just an example of how to retrieve goals from the new goal structure
+    const goals = new Map<number, AchievementGoal>();
+    // Converting the original goal array to a map is necessary because
+    // the goal index in the goals array might not be the same as its goal id
+    // Also, replace mockGoals with the goals from redux store
+    mockGoals.forEach(goal => goals.set(goal.id, goal));
```
Comment (anthony-halim) [inline src/features/game/state/GameUserStateManager.ts:97]: Just replace with 2 TODO comments

File: src/commons/achievement/utils/AchievementInferencer.ts [hunk 8]
```diff
@@ -359,51 +381,72 @@ class AchievementInferencer {
     node.displayDeadline = descendantDeadlines.reduce(compareDeadlines, node.displayDeadline);
   }
 
+  private generateMaxExp(node: InferencerNode) {
```
Comment (anthony-halim) [inline src/commons/achievement/utils/AchievementInferencer.ts:384]: generateSumExp. name is misleading

File: src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx (hunk 1 above)
Comment (jetkan-yk) [reply src/commons/achievement/control/editorTools/editableUtils/goals/EditableAchievementGoal.tsx:5]: This is part of upcoming changes in the next PR (same as below)

File: src/commons/achievement/utils/AchievementInferencer.ts (hunk 8 above)
Comment (jetkan-yk) [reply src/commons/achievement/utils/AchievementInferencer.ts:384]: It is to generate the maxExp of an achievement. Naming wont be consistent if the function is called generateSumExp while we have getMaxExp and the attribute is named maxExp.

**Pull request merged**